src/smtencoding/FStar.SMTEncoding.Encode.fs

```diff
@@ -343,6 +343,42 @@ let is_eta env vars t =
         | _ -> None in
   aux t (List.rev vars)
 
+(* [is_reifiable_* env x] returns true if the effect name/computational effect (of *)
+(* a body or codomain of an arrow) [x] is reifiable *)
+
+let is_reifiable_effect (env:Env.env) (effect_lid:lident) : bool =
+  let quals = Env.lookup_effect_quals env effect_lid in
+  List.contains Reifiable quals
+
+let is_reifiable (env:Env.env) (c:either<S.lcomp, S.residual_comp>) : bool =
+  let effect_lid = match c with
+    | Inl lc -> lc.eff_name
+    | Inr (eff_name, _) -> eff_name
+  in
+  is_reifiable_effect env effect_lid
+
+let is_reifiable_comp (env:Env.env) (c:S.comp) : bool =
+  match c.n with
+  | Comp ct -> is_reifiable_effect env ct.effect_name
+  | _ -> false
+
+let is_reifiable_function (env:Env.env) (t:S.term) : bool =
+  match (SS.compress t).n with
+  | Tm_arrow (_, c) -> is_reifiable_comp env c
+  | _ -> false
+
+(* [reify_body env t] assumes that [t] has a reifiable computation type *)
+(* that is env |- t : M t' for some effect M and type t' where M is reifiable *)
+(* and returns the result of reifying t *)
+let reify_body (env:Env.env) (t:S.term) : S.term =
+  let tm = U.mk_reify t in
+  let tm' = N.normalize [N.Beta; N.Reify; N.Eager_unfolding; N.EraseUniverses; N.AllowUnboundUniverses] env tm in
+  if Env.debug env <| Options.Other "SMTEncodingReify"
+  then BU.print2 "Reified body %s \nto %s\n"
+                (Print.term_to_string tm)
+                (Print.term_to_string tm') ;
+  tm'
+
```

Comment:
All of these seem generally useful, even outside the SMT encoding. Consider moving them to FStar.TypeChecker.Util, e.g.